### PR TITLE
Improve ListHeader

### PR DESCRIPTION
**Pull request by Isira-Seneviratne**

### CODE REVIEW

File: app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt [hunk 1]
```diff
@@ -4,32 +4,14 @@ import android.content.Context
 import androidx.annotation.StringRes
 import org.koitharu.kotatsu.core.ui.model.DateTimeAgo
 
-class ListHeader private constructor(
-	private val text: CharSequence?,
-	@StringRes private val textRes: Int,
-	private val dateTimeAgo: DateTimeAgo?,
-	@StringRes val buttonTextRes: Int,
-	val payload: Any?,
+data class ListHeader(
```
Comment (Koitharu) [inline app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt:7]: We avoid using data classes whenever possible on purpose. For data class compiler generate also `componentN()` and `copy` functions which are not needed but might increase apk size. Long story short, we use data classes only if we need `copy` function

Comment (Koitharu) [reply app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt:7]: This constructor also allow to pass, for instance, both `text` and `textRes`, or do not pass any arguments, that will produce invalid state

Comment (Isira-Seneviratne) [reply app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt:7]: The extra functions are removed by R8 during code shrinking if they aren't used.

Comment (Koitharu) [reply app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt:7]: Hmm, I guess it makes sense

Comment (Isira-Seneviratne) [reply app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt:7]: > This constructor also allow to pass, for instance, both `text` and `textRes`, or do not pass any arguments, that will produce invalid state

In that case, I'll remove the default arguments.

Comment (Koitharu) [reply app/src/main/kotlin/org/koitharu/kotatsu/list/ui/model/ListHeader.kt:7]: I'll try to replace some classes with `equals`/`hashCode` with `data`  and compare apk size

**Comment by Koitharu** [comment]
So, I converted 10 classes to `data` and removed manually generated `equals`/`hashCode`
Before conversion:
изображение
With data classes:
изображение
The difference is 10 methods
Looks like `copy()` and `componentN()` functions are absent, but `toString()` is present
изображение

So, I think, it is ok to use `data` classes (ofc, it may consume a bit more time at compilation stage, but this is insignificant), thank you for this

File: app/src/main/kotlin/org/koitharu/kotatsu/tracker/ui/feed/FeedViewModel.kt [hunk 2]
```diff
@@ -85,7 +85,7 @@ class FeedViewModel @Inject constructor(
 		for (item in this) {
 			val date = timeAgo(item.createdAt)
 			if (prevDate != date) {
-				destination += ListHeader(date, 0, null)
+				destination += ListHeader(null, 0, date)
```
Comment (Koitharu) [inline app/src/main/kotlin/org/koitharu/kotatsu/tracker/ui/feed/FeedViewModel.kt:88]: Now user **must** pass both `text` and `textRes`, however this is not 100% correct too:
- If user pass not-null `text` and 0 `textRes`, `text` will be used - ok
- Both not-null `text` and non-zero `textRes` -- what result is expected?
- ...

Summarize, I suppose solution with multiple constructors is the best (except `sealed class`, but this solution is too expensive for such small class`), so class might be `data` with private primary constructor and some public secondary constructors

**Review by Koitharu — APPROVED** [review]

**Pull request merged**